pkg/health/etcdmember/check_ready.go

```diff
@@ -41,48 +48,96 @@ func (r *readyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) []Resul
+		r.logger.Error(err, "failed to get leases for etcd member readiness check")
+	}
+
+	for _, lease := range leases.Items {
+		leaseDurationSeconds := lease.Spec.LeaseDurationSeconds
+		if leaseDurationSeconds == nil {
+			r.logger.Error(fmt.Errorf("leaseDurationSeconds not set for lease object %s/%s", lease.Namespace, lease.Name), "Failed to perform member readiness check")
 			continue
 		}
 
-		// Skip if status is not already Unknown and LastUpdateTime is within grace period.
-		if !member.LastUpdateTime.Time.Add(r.memberConfig.EtcdMemberUnknownThreshold).Before(checkTime) {
+		var (
+			id, role = separateIdFromRole(lease.Spec.HolderIdentity)
+			res      = &result{
+				id:   id,
+				name: lease.Name,
+				role: role,
+			}
+		)
+
+		// Check if member is in bootstrapping phase
+		// Members are supposed to be added to the members array only if they have joined the cluster (== RenewTime is set).
+		// This behavior is expected by the `Ready` condition and it will become imprecise if members are added here too early.
+		renew := lease.Spec.RenewTime
+		if renew == nil {
+			r.logger.Info("Member hasn't acquired lease yet, still in bootstrapping phase", "name", lease.Name)
 			continue
 		}
 
-		// If pod is not running or cannot be found then we deduce that the status is NotReady.
-		ready, err := r.checkContainersAreReady(ctx, etcd.Namespace, member)
-		if (err == nil && !ready) || apierrors.IsNotFound(err) {
-			results = append(results, &result{
-				id:     member.ID,
-				status: druidv1alpha1.EtcdMemeberStatusNotReady,
-				reason: "ContainersNotReady",
-			})
+		// Check if member state must be considered as not ready
+		if renew.Add(time.Duration(*leaseDurationSeconds) * time.Second).Add(r.memberConfig.EtcdMemberNotReadyThreshold).Before(checkTime) {
+			res.status = druidv1alpha1.EtcdMemeberStatusNotReady
+			res.reason = "LeaseExpired"
+			results = append(results, res)
 			continue
 		}
 
-		// For every other reason the status is Unknown.
-		results = append(results, &result{
-			id:     member.ID,
-			status: druidv1alpha1.EtcdMemeberStatusUnknown,
-			reason: "UnknownMemberStatus",
-		})
+		// Check if member state must be considered as unknown
+		if renew.Add(time.Duration(*leaseDurationSeconds) * time.Second).Before(checkTime) {
+			// If pod is not running or cannot be found then we deduce that the status is NotReady.
+			ready, err := r.checkContainersAreReady(ctx, lease.Namespace, lease.Name)
+			if (err == nil && !ready) || apierrors.IsNotFound(err) {
+				res.status = druidv1alpha1.EtcdMemeberStatusNotReady
+				res.reason = "ContainersNotReady"
+				results = append(results, res)
+				continue
+			}
+
+			res.status = druidv1alpha1.EtcdMemeberStatusUnknown
+			res.reason = "LeaseExpired"
+			results = append(results, res)
+			continue
+		}
+
+		res.status = druidv1alpha1.EtcdMemeberStatusReady
```

Comment:
Typo in the constants `EtcdMemeber...` should be `EtcdMember...`. Sorry, I didn't catch it the first time around.